Source/cmServerProtocol.cxx

```diff
@@ -815,6 +816,109 @@ static Json::Value DumpSourceFilesList(
   return result;
 }
 
+struct CaseInsensitiveHash
+{
+  size_t operator()(const std::string& keyVal) const
+  {
+    return std::hash<std::string>()(cmsys::SystemTools::LowerCase(keyVal));
+  }
+};
+
+struct CaseInsensitiveEqual
+{
+  bool operator()(const std::string& left, const std::string& right) const
+  {
+    return cmsys::SystemTools::LowerCase(left) == cmsys::SystemTools::LowerCase(right);
+  }
+};
+
+static Json::Value DumpSourceFilesListForInterfaceTarget(
+  cmGeneratorTarget* target,
+  const std::string& config)
+{
+  std::unordered_set<std::string, CaseInsensitiveHash, CaseInsensitiveEqual>
+    sources;
+  
+  cmLocalGenerator* lg = target->GetLocalGenerator();
+  // Leave language unspecified in the evaluator. If it's ever needed we
+  // have to compute it for each file based on its file type.
+  cmGeneratorExpressionInterpreter genexInterpreter(
+    lg, target, config, target->GetName(), "");
+  
+  // First see if there are any INTERFACE_SOURCES defined for this target.
+  auto targetProp = target->Target->GetProperty("INTERFACE_SOURCES");
+  if (targetProp != nullptr) {
+    const char* evaluatedSources =
+      genexInterpreter.Evaluate(targetProp, "INTERFACE_SOURCES");
+    auto components = cmsys::SystemTools::SplitString(evaluatedSources, ';');
+    if (!components.empty()) {
+      sources.insert(components.begin(), components.end());
+    }
+  }
+
+  // Next pick up sources from INTERFACE_INCLUDE_DIRECTORIES and
+  // INTERFACE_SYSTEM_INCLUDE_DIRECTORIES if defined for this target.
+  std::unordered_set<std::string> includeDirs;
+  targetProp = target->Target->GetProperty("INTERFACE_INCLUDE_DIRECTORIES");
+  if (targetProp != nullptr) {
+    const char* evaluatedIncludes =
+      genexInterpreter.Evaluate(targetProp, "INTERFACE_INCLUDE_DIRECTORIES");
+    auto components = cmsys::SystemTools::SplitString(evaluatedIncludes, ';');
+    if (!components.empty()) {
+      includeDirs.insert(components.begin(), components.end());
+    }
+  }
+  targetProp = target->Target->GetProperty("INTERFACE_SYSTEM_INCLUDE_DIRECTORIES");
+  if (targetProp != nullptr) {
+    const char* evaluatedIncludes =
+      genexInterpreter.Evaluate(targetProp, "INTERFACE_SYSTEM_INCLUDE_DIRECTORIES");
+    auto components = cmsys::SystemTools::SplitString(evaluatedIncludes, ';');
+    if (!components.empty()) {
+      includeDirs.insert(components.begin(), components.end());
+    }
+  }
+  
+  if (!includeDirs.empty()) {
```

Comment:
>includeDirs [](start = 7, length = 11)

Not sure about this part. It's enumerating all directories. This could take a long time. And what would it show if this was a remote machine in say the logical view? 

Can we just mark these results as dirs and then attempt to enumerate in the logical view as appropriate?

Reply:
You mean if the user clicks on it in the logical view, then we do a directory walk on the VS side? So I'd add a new type to the response to indicate this is a directory for an interface target?

Reply:
Yeah I think so. At least then we'd skip walking whole hard drives here.

---
In reply to: [196262746](https://github.com/Microsoft/CMake/pull/61#discussion_r196262746) [](ancestors = 196262746)

Reply:
Or just check on the other side if the type of the thing is a file or a directory. Probably don't even need a type difference here.

---
In reply to: [196262874](https://github.com/Microsoft/CMake/pull/61#discussion_r196262874) [](ancestors = 196262874,196262746)